exec/memory/base/src/main/java/io/netty/buffer/DrillBuf.java

```diff
@@ -703,7 +703,18 @@ protected void _setLong(int index, long value) {
 
   @Override
   public ByteBuf getBytes(int index, ByteBuf dst, int dstIndex, int length) {
-    udle.getBytes(index + offset, dst, dstIndex, length);
+    final int BULK_COPY_THR = 1024;
```

Comment:
The optimization is not obvious (and is counterintuitive) without more details.

Reply:
You are right, I'll put more information on this optimization:
- During code profiling, I have noticed that getBytes() doesn't perform well when called with small length (lower than 1k)
- Its throughput improves as the length increases

Analysis :
- Java exposes two intrinsics for writing to direct memory: putByte and copyMemory
- The JVM is able to inline memory access (no function call) for putByte 
- copyMemory is a bulk API and this internally invokes libc memcpy (requires function call)
- The rational is that we are willing to incur a function call if the associated processing is larger than the overhead; this is almost never the case for small memory accesses.

Reply:
My testing shows that `PlatformDependent.copyMemory` outperforms multiple calls to `PlatformDependent.getByte()`/`PlatformDependent.putByte()` starting with 16 bytes. Copying 1024 bytes buffer using `getByte()`/`putByte()` is 10x slower compared to `copyMemory()`.

Reply:
Vlad,

- I had a chat with @bitblender and he explains that Java was invoking a stub (not a function call) to perform copyMemory; he agreed copyMemory will be slower for small buffers and the task was to determine the cutoff point
- My tests (I will send you my test) indicate that a length of 1024bytes is the length were copyMemory starts performing exactly as getByte()

NOTE - I am using JRE 1.8; static buffers initialized once; payload 1MB (1048576bytes) and loop-count of 102400; MacOS High Sierra; 1 thread, 4GB MX, MS